chain/solana/src/data_source.rs

```diff
@@ -0,0 +1,406 @@
+            ))
+        }
+
+        // Validate that there is a `source` address if there are instruction handlers
+        let no_source_address = self.address().is_none();
+        let has_instruction_handlers = !self.mapping.instruction_handlers.is_empty();
+        if no_source_address && has_instruction_handlers {
+            errors.push(SubgraphManifestValidationError::SourceAddressRequired.into());
+        };
+
+        // Validate that there are no more than one of both block handlers and receipt handlers
+        if self.mapping.block_handlers.len() > 1 {
+            errors.push(anyhow!("data source has duplicated block handlers"));
+        }
+        if self.mapping.instruction_handlers.len() > 1 {
+            errors.push(anyhow!("data source has duplicated receipt handlers"));
+        }
+
+        errors
+    }
+}
+
+impl DataSource {
+    fn from_manifest(
+        kind: String,
+        network: Option<String>,
+        name: String,
+        source: Source,
+        mapping: Mapping,
+        context: Option<DataSourceContext>,
+    ) -> Result<Self, Error> {
+        // Data sources in the manifest are created "before genesis" so they have no creation block.
+        let creation_block = None;
+
+        Ok(DataSource {
+            kind,
+            network,
+            name,
+            source,
+            mapping,
+            context: Arc::new(context),
+            creation_block,
+        })
+    }
+
+    fn handler_for_block(&self) -> Option<&MappingBlockHandler> {
+        self.mapping.block_handlers.first()
+    }
+
+    fn handler_for_instruction(&self) -> Option<&MappingInstructionHandler> {
+        self.mapping.instruction_handlers.first()
+    }
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, Deserialize)]
+pub struct UnresolvedDataSource {
+    pub kind: String,
+    pub network: Option<String>,
+    pub name: String,
+    pub(crate) source: Source,
+    pub mapping: UnresolvedMapping,
+    pub context: Option<DataSourceContext>,
+}
+
+#[async_trait]
+impl blockchain::UnresolvedDataSource<Chain> for UnresolvedDataSource {
+    async fn resolve(
+        self,
+        resolver: &impl LinkResolver,
+        logger: &Logger,
+    ) -> Result<DataSource, Error> {
+        let UnresolvedDataSource {
+            kind,
+            network,
+            name,
+            source,
+            mapping,
+            context,
+        } = self;
+
+        info!(logger, "Resolve data source"; "name" => &name, "source" => &source.start_block);
+
+        let mapping = mapping.resolve(&*resolver, logger).await?;
+
+        DataSource::from_manifest(kind, network, name, source, mapping, context)
+    }
+}
+
+impl TryFrom<DataSourceTemplateInfo<Chain>> for DataSource {
+    type Error = Error;
+
+    fn try_from(_info: DataSourceTemplateInfo<Chain>) -> Result<Self, Error> {
+        Err(anyhow!("Near subgraphs do not support templates"))
+
+        // How this might be implemented if/when Near gets support for templates:
```

Comment:
I think we can remove it and later when we will need to implemented it, check NEAR implementation (which is now tracked in #3202